src/emulator/storage/metadata.ts

```diff
@@ -30,35 +39,51 @@ export class StoredFileMetadata {
   customMetadata: { [s: string]: string };
 
   constructor(
-    bucketId: string,
-    objectId: string,
     bytes: Buffer,
-    contentType: string,
-    contentEncoding: string | undefined,
-    incomingMetadata: IncomingMetadata,
+    opts: Partial<SerializedFileMetadata> & {
+      name: string;
+      bucket: string;
+      contentType: string;
+    },
+    incomingMetadata: IncomingMetadata | undefined,
     private _cloudFunctions: StorageCloudFunctions
   ) {
-    this.name = objectId;
-    this.bucket = bucketId;
-    this.timeCreated = new Date();
-    this.size = bytes.byteLength;
-    this.metageneration = 1;
-    this.generation = Date.now();
-    this.md5Hash = generateMd5Hash(bytes);
-    this.storageClass = "STANDARD";
-    this.downloadTokens = "";
-    this.etag = "someETag";
-    this.crc32c = `${crc32c(bytes)}`;
-    this.contentDisposition = "inline";
-    this.updated = this.timeCreated;
-    this.contentType = contentType;
-    this.cacheControl = "no-cache";
-    this.contentLanguage = "en-us";
-    this.contentEncoding = contentEncoding || "identity";
-    this.customMetadata = incomingMetadata.metadata || {};
-
-    this.addDownloadToken();
-    this.update(incomingMetadata);
+    // Required fields
+    this.name = opts.name;
+    this.bucket = opts.bucket;
+    this.contentType = opts.contentType;
+
+    // Optional fields
+    this.metageneration = opts.metageneration || 1;
+    this.generation = opts.generation || Date.now();
+    this.storageClass = opts.storageClass || "STANDARD";
+    this.etag = opts.etag || "someETag";
+    this.contentDisposition = opts.contentDisposition || "inline";
+    this.cacheControl = opts.cacheControl || "no-cache";
+    this.contentLanguage = opts.contentLanguage || "en-us";
+    this.contentEncoding = opts.contentEncoding || "identity";
+    this.customMetadata = opts.customMetadata || {};
+
+    // Special handling for date fields
+    this.timeCreated = opts.timeCreated ? new Date(opts.timeCreated) : new Date();
+    this.updated = opts.updated ? new Date(opts.updated) : this.timeCreated;
+
+    // Fields derived from bytes
+    this.size = opts.size || bytes.byteLength;
```

Comment:
We should probably extract these using a helper function outside of this constructor (to avoid having to pass the raw/dummy buffer (data) into Metadata)